autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/imputation/SimpleImputer.py

```diff
@@ -15,92 +13,142 @@
-                                                                        copy=False)
+                # Train data is numpy as of this point, where an Ordinal Encoding is used
+                # for categoricals. Only Numbers are allowed for `fill_value`
+                imputer = SklearnSimpleImputer(strategy='constant', fill_value=-1, copy=False)
+                self.preprocessor['categorical'] = imputer
             else:
-                self.preprocessor['categorical'] = SklearnSimpleImputer(strategy=self.categorical_strategy,
-                                                                        copy=False)
-        numerical_columns = X['dataset_properties']['numerical_columns'] \
-            if isinstance(X['dataset_properties']['numerical_columns'], List) else []
-        if len(numerical_columns) != 0:
+                imputer = SklearnSimpleImputer(strategy=self.categorical_strategy, copy=False)
+                self.preprocessor['categorical'] = imputer
+
+        # Choose an imputer for any numerical columns
+        numerical_columns = X['dataset_properties']['numerical_columns']
+
+        if isinstance(numerical_columns, List) and len(numerical_columns) > 0:
             if self.numerical_strategy == 'constant_zero':
-                self.preprocessor['numerical'] = SklearnSimpleImputer(strategy='constant',
-                                                                      fill_value=0,
-                                                                      copy=False)
+                imputer = SklearnSimpleImputer(strategy='constant', fill_value=0, copy=False)
+                self.preprocessor['numerical'] = imputer
             else:
-                self.preprocessor['numerical'] = SklearnSimpleImputer(strategy=self.numerical_strategy, copy=False)
+                imputer = SklearnSimpleImputer(strategy=self.numerical_strategy, copy=False)
+                self.preprocessor['numerical'] = imputer
 
         return self
 
     @staticmethod
     def get_hyperparameter_search_space(
         dataset_properties: Optional[Dict[str, BaseDatasetPropertiesType]] = None,
-        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(hyperparameter='numerical_strategy',
-                                                                                  value_range=("mean", "median",
-                                                                                               "most_frequent",
-                                                                                               "constant_zero"),
-                                                                                  default_value="mean",
-                                                                                  ),
+        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(
+            hyperparameter='numerical_strategy',
+            value_range=("mean", "median", "most_frequent", "constant_zero"),
```

Comment:
It would be better if we can use `Enum`.
We did not have it so much just because I did not join the refactoring phase.
At least, I would like to reduce so many dependencies on hard-coded strings as much as possible.
Because this file also uses so many hard-coded strings, which we can avoid by enum.

```
from enum import Enum
from typing import List


class NumericalImputerChoices(Enum):
    mean = "mean"
    median = "median"
    most_frequent = "most_frequent"
    constant_zero = "constant_zero"

    @classmethod
    def get_choices(cls) -> List[str]:
        return [c.value for c in cls]


class CategoricalImputerChoices(Enum):
    most_frequent = "most_frequent"
    constant_missing = "constant_!missing!"

    @classmethod
    def get_choices(cls) -> List[str]:
        return [c.value for c in cls]
```

Reply:
I can do that, you do however require to update these whenever sklearn updates them as well as that is where they are forwarded to.

Reply:
I can see why you want to do that but it makes the code that uses it a little less pretty. For example, it's hard to see where the error with the follow code is:

```python
        numerical_strategy: HyperparameterSearchSpace = HyperparameterSearchSpace(
            hyperparameter='numerical_strategy',
            value_range=NumericalImputerChoice.get_choices(),
            default_value=NumericalImputerChoice.mean,
        ),
```

It should be `default_value=NumericalImputerChoice.mean.value`. The problem here is that Enum's are namespaces `NumericalImputerChoice.mean == "NumericalImputerChoice.mean"` where as we would like `NumericalImputerChoice.mean == "mean"`. This is more readily achievable with a `NamedTuple`. I will implement the Enum version and you can change it if you like or leave it as is

Reply:
Another clean'ish solution is just ditch the enum part. 

```python
class Choices:
    x: str = "x"
    y: str = "y"
    
assert Choices.x = "x"
```

The type is still a string, meaning it's easy to use, people don't need to know about the existence of this class to use it, they can still pass a string. It also allows for the internal code to use the class, where we do know about it.

Reply:
As another note, it also makes parameters extremely long, it's 101 characters which is over the character count limit of the checker:
```python
    def __init__(
        self,
        random_state: Optional[np.random.RandomState] = None,
        numerical_strategy: NumericalImputerChoice = NumericalImputerChoice.mean.value,
        categorical_strategy: CategoricalImputerChoice = CategoricalImputerChoice.most_frequent.value
    ):
```

Reply:
I've not changed it for now, there's too many decisions to make that I think can be addressed based on how you would like it done yourself. I've changed it back to strings.

Reply:
Hey, thanks for the insights. Yeah, I think such changes should be a part of a separate PR as this PR is meant to clean up the messy statements in this file.

Reply:
All the problems come from the fact that we still need to follow python3.7.
Once we switched to python3.8, we can use `Literal`...

Reply:
We have the same thing, there is `Literal` at the moment but the type checking for it doesn't work as expected.

```python
def f(s: Literal['yes', 'no'])
   ...
   
f("yes") # Type Error

param: Literal = "yes"
f(param) # Okay
```

Reply:
All the problems come from the fact that we still need to follow python3.7.
Once we switched to python3.8, we can use `Literal`...

> in general, enum values should never be directly used, such as their string value

Yeah, I agree with you. I am just uncomfortable using software that assumes that users google string choices (typically sklearn).
And I found it really useful to be able to use `enum` or `class` such as in numpy, facebook ax..
But I got your point, let's stay for now.
I did not know some points you mentioned, thanks for raising them. @eddiebergman .

(just a question, but) the better solution will be something like this:
```
class NumericalImputerChoices(NamedTuple):
    mean = "mean"
    median = "median"

num_imputer_choices = NumericalImputerChoices()
```
and use `str` for all the typings.

Btw, you do not need to put `.value` (FYI).

Reply:
also, you can add `python` after the backticks ``` of a code block to get python highlighting ;)